pkg/auth/providers/azuread/azuread_provider.go

```diff
@@ -0,0 +1,294 @@
+			principals = append(principals, p)
+		}
+	}
+
+	for _, principal := range principals {
+		if principal.PrincipalType == userType {
+			if p.isThisUserMe(token.UserPrincipal, principal) {
+				principal.Me = true
+			}
+		} else if principal.PrincipalType == groupType {
+			if p.isMemberOf(token.GroupPrincipals, principal) {
+				principal.MemberOf = true
+			}
+		}
+	}
+
+	return principals, nil
+}
+
+func (p *azureADProvider) GetPrincipal(principalID string, token v3.Token) (v3.Principal, error) {
+	config, err := p.getAzureADConfig()
+	if err != nil {
+		return v3.Principal{}, nil
+	}
+	accessToken := token.ProviderInfo["access_token"]
+
+	var externalID string
+	parts := strings.SplitN(principalID, ":", 2)
+	if len(parts) != 2 {
+		return v3.Principal{}, errors.Errorf("invalid id %v", principalID)
+	}
+	externalID = strings.TrimPrefix(parts[1], "//")
+	parts = strings.SplitN(parts[0], "_", 2)
+	if len(parts) != 2 {
+		return v3.Principal{}, errors.Errorf("invalid id %v", principalID)
+	}
+
+	principalType := parts[1]
+	var acct Account
+	switch principalType {
+	case userType:
+		acct, err = p.azureADClient.getUserByID(externalID, accessToken, config)
+		if err != nil {
+			return v3.Principal{}, err
+		}
+	case groupType:
+		acct, err = p.azureADClient.getGroupByID(externalID, accessToken, config)
+		if err != nil {
+			return v3.Principal{}, err
+		}
+	default:
+		return v3.Principal{}, fmt.Errorf("Cannot get the azuread account due to invalid externalIDType %v", principalType)
+	}
+
+	princ := p.toPrincipal(principalType, acct, &token)
+	return princ, nil
+}
+
+func (p *azureADProvider) isThisUserMe(me v3.Principal, other v3.Principal) bool {
+	if me.ObjectMeta.Name == other.ObjectMeta.Name && me.LoginName == other.LoginName && me.PrincipalType == other.PrincipalType {
+		return true
+	}
+	return false
+}
+
+func (p *azureADProvider) isMemberOf(myGroups []v3.Principal, other v3.Principal) bool {
+	for _, mygroup := range myGroups {
+		if mygroup.ObjectMeta.Name == other.ObjectMeta.Name && mygroup.PrincipalType == other.PrincipalType {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *azureADProvider) getAzureADConfig() (*v3.AzureADConfig, error) {
+	authConfigObj, err := p.authConfigs.ObjectClient().UnstructuredClient().Get(Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve AzureADConfig, error: %v", err)
+	}
+
+	u, ok := authConfigObj.(runtime.Unstructured)
+	if !ok {
+		return nil, fmt.Errorf("failed to retrieve AzureADConfig, cannot read k8s Unstructured data")
+	}
+	storedAzureADConfigMap := u.UnstructuredContent()
+
+	storedAzureADConfig := &v3.AzureADConfig{}
+	mapstructure.Decode(storedAzureADConfigMap, storedAzureADConfig)
+
+	metadataMap, ok := storedAzureADConfigMap["metadata"].(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("failed to retrieve AzureADConfig metadata, cannot read k8s Unstructured data")
+	}
+
+	typemeta := &metav1.ObjectMeta{}
```

Comment:
typeMeta is a different field on k8s resources. it is confusing to name this variable that. can you change the name objectMeta or something like that?